quinn/src/connection.rs

```diff
@@ -26,18 +23,60 @@ pub use crate::tls::{Certificate, CertificateChain, PrivateKey};
 pub use crate::builders::{
     ClientConfigBuilder, EndpointBuilder, EndpointError, ServerConfigBuilder,
 };
+use crate::streams::{NewStream, RecvStream, SendStream};
 use crate::{ConnectionEvent, EndpointEvent};
 
 /// Connecting future
-pub struct ConnectingFuture(Option<ConnectionDriver>);
+pub struct Connecting(Option<ConnectionDriver>);
 
-impl ConnectingFuture {
+impl Connecting {
     pub(crate) fn new(conn: ConnectionRef) -> Self {
         Self(Some(ConnectionDriver(conn)))
     }
+
+    /// Convert into a 0-RTT or 0.5-RTT connection at the cost of weakened security
+    ///
+    /// Opens up the connection for use before the handshake finishes, allowing the API user to
+    /// send data with 0-RTT encryption if the necessary key material is available. This is useful
+    /// for reducing start-up latency by beginning transmission of application data without waiting
+    /// for the handshake's cryptographic security guarantees to be established.
+    ///
+    /// # Security
+    ///
+    /// On outgoing connections, this enables transmisison of 0-RTT data, which might be vulnerable
+    /// to replay attacks, and should therefore never invoke non-idempotent operations.
+    ///
+    /// On incoming connections, this enables transmission of 0.5-RTT data, which might be
+    /// intercepted by a man-in-the-middle. If this occurs, the handshake will not complete
+    /// successfully.
+    ///
+    /// Once [`Connection::is_handshaking`] returns false, further transmissions are unaffected by
+    /// these weaknesses.
+    ///
+    /// # Errors
+    ///
+    /// Outgoing connections are only 0-RTT-capable when a cryptographic session ticket cached from
+    /// a previous connection to the same server is available, and includes a 0-RTT key. If no such
+    /// ticket is found, `self` is returned unmodified.
+    ///
+    /// For incoming connections, a 0.5-RTT connection will always be successfully constructed.
+    pub fn into_0rtt(mut self) -> Result<(ConnectionDriver, Connection, IncomingStreams), Self> {
+        if (self.0.as_mut().unwrap().0)
+            .lock()
+            .unwrap()
+            .inner
+            .has_0rtt()
+        {
+            let ConnectionDriver(conn) = self.0.take().unwrap();
+            conn.lock().unwrap().driver.take();
```

Comment:
Didn't you decide to remove this?

Reply:
Yeah, not sure what happened with that, fixed now though!